### DISCUSSION TITLE
Additional claims per application?

### DISCUSSION THREAD

**Discussion by dreamair** [zitadel/zitadel#8698] [Q&A]
Hi
I've managed to get a Minio instance to authenticate via Zitadel. But only with an action adding a "policy" claim to the ID token.
```
function minio(ctx, api) {
  api.v1.claims.setClaim('policy', 'readwrite')
}
```
While this works, I now have this claim in every other application, other project even, also included. That's ok for now, but according to murphy's law there will be another app requesting a "policy" claim to be set to "all-open" soon ;)
Is there a way to specify an additional claim, or an action to set that claim, for one application only?
For the use case list item "As a user I want to extend the claims of the id token to add information needed in my app" in [the future of actions](https://github.com/zitadel/zitadel/discussions/6926) I couldn't find any follow up or how-to.
Thanks!

**Comment by hifabienne** [comment]
Hi @dreamair 

I think you could check for the audience and only include the claim, if the audience is the app you are looking for.

**Reply by dreamair** [reply]
Hi @hifabienne 

Thanks for your answer. In my debugging I couldn't find the audience within an action. "ctx.v1.claims" seem to only include the user claims. Within an action of a "Complement Token Flow" that is. How do I set a claim within an action of "Internal/External Authentication Flow"?
Have I missed something?

**Reply by hifabienne** [reply]
@adlerhurst @stebenz can you help here?

**Reply by adlerhurst** [reply]
I think you can adapt this example: https://github.com/zitadel/actions/blob/main/examples/execute_action_on_specific_app.js

**Reply by dreamair** [reply]
Thanks @adlerhurst . But where do I set/add a claim in an action of a "Internal/External Authentication Flow"? and the name "Post Authentication" suggest it's after the authentication and the token is already out. Isn't the "Complement Token Flow" the place to hook into? ...and there the authRequest with the applicationId and audiences is missing 😒

**Reply by adlerhurst** [reply]
sorry for the late reply.

You are right. You are looking for the complement token flow. If you have use user grants to grant access to the user you could use [the ctx.v1.user.grants](https://zitadel.com/docs/apis/actions/complement-token#pre-userinfo-creation-id_token--userinfo--introspection-endpoint) to at least get the correct project id. sadly you currently cannot get the auth request in this flow.
